### PR TITLE
[BugFix] update args type when rewrite the input of agg funcs

### PR DESCRIPTION
**Pull request by packy92**
## What type of PR is this：
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

In `GroupByCountDistinctRewriteRule` we do things as below:
```
v1 is the distribution key
# before
select count(distinct v1), sum(v2) from t1 group by v3;

# after
select count(v1), sum(sum1) from 
     (select v1, sum(v2) sum1 from t1 group by v1, v3) group by v3;
```
After this transformation, we actually have two different agg operators. The top agg use the output from bottom agg as its input. Hence, we need update the type of argument in callOperator `sum(sum1)`.

Other changes:
We have only four implementation to transform a one stage agg to multiple stage agg. But we have more than ten branches to these implmentations. So I simplifies the code based on the old logic. Now the steps to choose a implementation are as follows:
1. check if it need multiple stage agg. If not, return empty.
2. check the number of distinct agg function. If it's zero, choose two stage agg.
3. check if these distinct agg function are supported. If not, throw exception.
4. check if it is suitable for two stage agg(the check logic is in `isSuitableForTwoStage()` method). If true, choose two stage agg.
5. check if it contains group by keys. 
    If ture: 
    - if all group by keys are constant, choose `implementOneDistinctWithConstantGroupByAgg`
    - if not, choose `implementOneDistinctWithGroupByAgg`
    
    else, choose `implementOneDistinctWithoutGroupByAgg`.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 81 / 95 (85.26%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/TableFunction.java|0|6|00.00%|[69, 70, 71, 72, 74, 140]|
|:large_blue_circle:|com/starrocks/catalog/ScalarFunction.java|0|6|00.00%|[96, 97, 98, 99, 100, 241]|
|:large_blue_circle:|com/starrocks/catalog/Function.java|19|20|95.00%|[776]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/SplitAggregateRule.java|31|32|96.88%|[148]|
|:large_blue_circle:|com/starrocks/catalog/AggregateFunction.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java|22|22|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 81 / 95 (85.26%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/TableFunction.java|0|6|00.00%|[69, 70, 71, 72, 74, 140]|
|:large_blue_circle:|com/starrocks/catalog/ScalarFunction.java|0|6|00.00%|[96, 97, 98, 99, 100, 241]|
|:large_blue_circle:|com/starrocks/catalog/Function.java|19|20|95.00%|[776]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/SplitAggregateRule.java|31|32|96.88%|[148]|
|:large_blue_circle:|com/starrocks/catalog/AggregateFunction.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java|22|22|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SplitAggregateRule.java [hunk 1]
```diff
@@ -128,42 +128,30 @@ private boolean needGenerateMultiStageAggregate(OptExpression input, List<CallOp
...
-            return false;
-        }
-        boolean hasSameMultiColumn = false;
+    // check if multi distinct functions used the same columns.
+    private void checkDistinctAgg(List<CallOperator> distinctAggCallOperator) {
         List<ScalarOperator> distinctChild0 = distinctAggCallOperator.get(0).getChildren();
         for (int i = 1; i < distinctAggCallOperator.size(); ++i) {
             List<ScalarOperator> distinctChildI = distinctAggCallOperator.get(i).getChildren();
             if (!distinctChild0.equals(distinctChildI)) {
-                if (distinctChild0.size() > 1 || distinctChildI.size() > 1) {
```
[the remark this answers is not in the document]
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SplitAggregateRule.java:152]: `RewriteMultiDistinctRule` will transform sql only contain one cloumn distinct agg like `count(distinct a), count(distinct b)` to `multi_distinct_count(a), multi_distinct_count(b)` which also removes their distinct flag. So I remove this line.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java [hunk 2]
```diff
@@ -198,19 +189,41 @@ public List<OptExpression> transform(OptExpression input, OptimizerContext conte
...
+                return origin;
+            } else {
+                Function newFunc = Expr.getBuiltinFunction(firstFuncName, origin.getFunction().getArgs(),
+                        Function.CompareMode.IS_NONSTRICT_SUPERTYPE_OF);
+                return new CallOperator(firstFuncName, newFunc.getReturnType(), args, newFunc);
+            }
+        } else {
+            String secondFuncName = OTHER_FUNCTION_TRANS.get(originFuncName).second;
+            Type[] argTypes = args.stream().map(ScalarOperator::getType).toArray(Type[]::new);
+            Function newFunc = origin.getFunction().updateArgType(argTypes);
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java:206]: update args type may lead to using the wrong method, it is better to use `Expr.getBuiltinFunction`?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SplitAggregateRule.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SplitAggregateRule.java:152]: Firstly, when analyze the statement, the distinct info will saved only in `count(distinct col)`, `sum(distinct col)`, other agg function's distinct info is meaningless and has been removed. Secondly, the `RewriteMultiDistinctRule` transform multiple distinct agg fucntions to `multi_distinct_count()` or `multi_distinct_sum()` and remove the `isDistinct` info. So in the `SplitAggregateRule`, there cannot exist scene like `select count(distinct a), count(distinct b)`.

[the remark this answers is not in the document]
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SplitAggregateRule.java:152]: In other words, if there exists distinctAgg in `SplitAggregateRule`, it must contains at least one multiple columns distinct agg fucntion.

**Review by kangkaisen — DISMISSED** [review]

**Review by Seaven — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by packy92** [comment]
run starrocks_fe_unittest

**Review by Seaven — APPROVED** [review]

**Comment by packy92** [comment]
run starrocks_admit_test

**Review by kangkaisen — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SplitAggregateRule.java [hunk 3]
```diff
@@ -118,8 +116,10 @@ private boolean mustGenerateMultiStageAggregate(OptExpression input, List<CallOp
 
     // Note: This method logic must consistent with CostEstimator::needGenerateOneStageAggNode
     private boolean needGenerateMultiStageAggregate(OptExpression input, List<CallOperator> distinctAggCallOperator) {
-        // 1. check if can generate multi stage aggregate.
-        if (!canGenerateMultiStageAggregate(input)) {
+        // 1. Must do one stage aggregate If the child contains limit,
+        //    the aggregation must be a single node to ensure correctness.
+        //    eg. select count(*) from (select * table limit 2) t
+        if (input.inputAt(0).getOp().hasLimit()) {
```
Comment (liuyehcf) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SplitAggregateRule.java:122]: is it possible to have child without limit but grandchild with limit?

**Review by liuyehcf — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SplitAggregateRule.java (hunk 3 above)
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SplitAggregateRule.java:122]: The rule is for  agg->leaf pattern. I think we only consider child is OK.

**Comment by packy92** [comment]
run starrocks_fe_unittest

**Comment by packy92** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 79 / 95 (83.16%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/TableFunction.java|0|6|00.00%|[69, 70, 71, 72, 74, 140]|
|:large_blue_circle:|com/starrocks/catalog/ScalarFunction.java|0|6|00.00%|[96, 97, 98, 99, 100, 241]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java|20|22|90.91%|[197, 199]|
|:large_blue_circle:|com/starrocks/catalog/Function.java|19|20|95.00%|[776]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/SplitAggregateRule.java|31|32|96.88%|[148]|
|:large_blue_circle:|com/starrocks/catalog/AggregateFunction.java|9|9|100.00%|[]|

**Comment by packy92** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.5

**Pull request merged**